File: app/src/test/java/com/mxt/anitrend/util/MediaUtilTests.java
package com.mxt.anitrend.util;

import com.mxt.anitrend.model.entity.anilist.MediaList;
import com.mxt.anitrend.model.entity.anilist.meta.MediaTitle;
import com.mxt.anitrend.model.entity.anilist.meta.MediaTrend;
import com.mxt.anitrend.model.entity.base.MediaBase;
import com.mxt.anitrend.util.media.MediaUtil;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.mxt.anitrend.util.KeyUtil.ANIME;
import static com.mxt.anitrend.util.KeyUtil.MANGA;
import static com.mxt.anitrend.util.KeyUtil.MediaStatus;
import static com.mxt.anitrend.util.KeyUtil.NOT_YET_RELEASED;
import static com.mxt.anitrend.util.KeyUtil.RELEASING;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.arrayContainingInAnyOrder;
import static org.hamcrest.Matchers.empty;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.StrictStubs.class)
public class MediaUtilTests {

    @Mock
    private MediaList list;
    @Mock
    private MediaBase media;
    @Mock
    private MediaTitle mediaTitle;

    @Before
    public void setupMocks() {
        when(list.getMedia()).thenReturn(media);
        when(media.getTitle()).thenReturn(mediaTitle);
    }

    @After
    public void resetMocks() {
        reset(list, media, mediaTitle);
    }

    /*
    isAnimeType
     */

    @Test
    public void isAnimeType_givenNull_shouldReturnFalse() {
        assertThat(MediaUtil.isAnimeType(null), is(false));
    }

    @Test
    public void isAnimeType_givenAnime_shouldReturnTrue() {
        when(media.getType()).thenReturn(ANIME);
        assertThat(MediaUtil.isAnimeType(media), is(true));
    }

    @Test
    public void isAnimeType_givenManga_shouldReturnFalse() {
        when(media.getType()).thenReturn(MANGA);
        assertThat(MediaUtil.isAnimeType(media), is(false));
    }

    /*
    isMangaType
     */

    @Test
    public void isMangaType_givenNull_shouldReturnFalse() {
        assertThat(MediaUtil.isMangaType(null), is(false));
    }

    @Test
    public void isMangaType_givenAnime_shouldReturnFalse() {
        when(media.getType()).thenReturn(ANIME);
        assertThat(MediaUtil.isMangaType(media), is(false));
    }

    @Test
    public void isMangaType_givenManga_shouldReturnTrue() {
        when(media.getType()).thenReturn(MANGA);
        assertThat(MediaUtil.isMangaType(media), is(true));
    }

    /*
    isIncrementLimitReached
     */

    @Test
    public void isIncrementLimitReached_ifProgressEqualToAnimeEpisodes_shouldReturnTrue() {
        int episodes = 10;
        //noinspection UnnecessaryLocalVariable
        int progress = episodes;

        when(media.getType()).thenReturn(ANIME);
        when(list.getProgress()).thenReturn(progress);
        when(media.getEpisodes()).thenReturn(episodes);

        assertThat(MediaUtil.isIncrementLimitReached(list), is(true));
    }

    @Test
    public void isIncrementLimitReached_ifProgressLessThanAnimeEpisodes_shouldReturnFalse() {
        int episodes = 10;
        int progress = 3;

        when(media.getType()).thenReturn(ANIME);
        when(list.getProgress()).thenReturn(progress);
        when(media.getEpisodes()).thenReturn(episodes);

        assertThat(MediaUtil.isIncrementLimitReached(list), is(false));
    }

    @Test
    public void isIncrementLimitReached_ifProgressGreaterThanAnimeEpisodes_shouldReturnFalse() {
        int episodes = 10;
        int progress = 15;

        when(media.getType()).thenReturn(ANIME);
        when(list.getProgress()).thenReturn(progress);
        when(media.getEpisodes()).thenReturn(episodes);

        assertThat(MediaUtil.isIncrementLimitReached(list), is(false));
    }

    @Test
    public void isIncrementLimitReached_ifProgressEqualToMangaChapters_shouldReturnTrue() {
        int chapters = 10;
        //noinspection UnnecessaryLocalVariable
        int progress = chapters;

        when(media.getType()).thenReturn(MANGA);
        when(list.getProgress()).thenReturn(progress);
        when(media.getChapters()).thenReturn(chapters);

        assertThat(MediaUtil.isIncrementLimitReached(list), is(true));
    }

    @Test
    public void isIncrementLimitReached_ifProgressLessThanMangaChapters_shouldReturnFalse() {
        int chapters = 10;
        int progress = 7;

        when(media.getType()).thenReturn(MANGA);
        when(list.getProgress()).thenReturn(progress);
        when(media.getChapters()).thenReturn(chapters);

        assertThat(MediaUtil.isIncrementLimitReached(list), is(false));
    }

    @Test
    public void isIncrementLimitReached_ifProgressGreaterThanMangaChapters_shouldReturnFalse() {
        int chapters = 10;
        //noinspection UnnecessaryLocalVariable
        int progress = 20;

        when(media.getType()).thenReturn(MANGA);
        when(list.getProgress()).thenReturn(progress);
        when(media.getChapters()).thenReturn(chapters);

        assertThat(MediaUtil.isIncrementLimitReached(list), is(false));
    }

    /*
    isAllowedStatus
     */

    @Test
    public void isAllowedStatus_ifMediaIsNotYetReleased_shouldReturnFalse() {
        when(media.getStatus()).thenReturn(NOT_YET_RELEASED);
        assertThat(MediaUtil.isAllowedStatus(list), is(false));
    }

    @Test
    public void isAllowedStatus_forAnyOtherStatus_shouldReturnTrue() {
        for (String status : MediaStatus) {
            if (!NOT_YET_RELEASED.equals(status)) {
                when(media.getStatus()).thenReturn(status);
                assertThat("Incrementing should be allowed for status: " + status,
                        MediaUtil.isAllowedStatus(list), is(true));
            }
        }
    }

    /*
    getMediaTitle
     */

    @Test
    public void getMediaTitle_shouldReturnUserPreferredTitle() {
        String title = "Gintama";
        when(mediaTitle.getUserPreferred()).thenReturn(title);
        assertThat(MediaUtil.getMediaTitle(media), equalTo(title));
    }

    /*
    getMediaListTitle
     */

    @Test
    public void getMediaListTitle_shouldReturnUserPreferredTitle() {
        String title = "Gintama";
        when(mediaTitle.getUserPreferred()).thenReturn(title);
        assertThat(MediaUtil.getMediaListTitle(list), equalTo(title));
    }

    /*
    mapMediaTrend
     */

    @Test
    public void mapMediaTrend_shouldReturnCorrespondingMedia() {

        MediaBase media1 = mock(MediaBase.class);
        MediaBase media2 = mock(MediaBase.class);
        MediaBase media3 = mock(MediaBase.class);
        List<MediaBase> mediaList = Arrays.asList(media1, media2, media3);


        List<MediaTrend> trendList = mediaList.stream().map(media -> {
            MediaTrend trend = mock(MediaTrend.class);
            when(trend.getMedia()).thenReturn(media);
            return trend;
        }).collect(Collectors.toList());

        assertThat(MediaUtil.mapMediaTrend(trendList).toArray(),
                arrayContainingInAnyOrder(media1, media2, media3));
    }

    @Test
    public void mapMediaTrend_givenNull_shouldReturnEmptyList() {
        assertThat(MediaUtil.mapMediaTrend(null), empty());
    }

    /*
    getAiringMedia
     */

    @Test
    public void getAiringMedia_shouldReturnReleasingMediaOnly() {
        List<MediaList> releasing =
                Stream.of(
                        mock(MediaBase.class),
                        mock(MediaBase.class))
                .map(media -> {
                    MediaList list = mock(MediaList.class);
                    when(list.getMedia()).thenReturn(media);
                    return list;
                }).collect(Collectors.toList());

        releasing.forEach(media -> when(media.getMedia().getStatus()).thenReturn(RELEASING));

        List<MediaList> notReleasing =
                Stream.of(MediaStatus)
                        .filter(status -> !RELEASING.equals(status))
                        .map(status -> {
                            MediaBase media = mock(MediaBase.class);
                            when(media.getStatus()).thenReturn(status);
                            return media;
                        })
                        .map(media -> {
                            MediaList list = mock(MediaList.class);
                            when(list.getMedia()).thenReturn(media);
                            return list;
                        }).collect(Collectors.toList());

        List<MediaList> allMedia = new ArrayList<>(releasing);
        allMedia.addAll(notReleasing);

        assertThat(MediaUtil.getAiringMedia(allMedia).toArray(), equalTo(releasing.toArray()));
    }

    @Test
    public void getAiringMedia_givenNull_shouldReturnEmptyList() {
        assertThat(MediaUtil.getAiringMedia(null), empty());
    }
}

File: app/src/test/java/com/mxt/anitrend/util/MediaUtilTests.kt
package com.mxt.anitrend.util

import com.mxt.anitrend.model.entity.anilist.MediaList
import com.mxt.anitrend.model.entity.anilist.meta.MediaTitle
import com.mxt.anitrend.model.entity.anilist.meta.MediaTrend
import com.mxt.anitrend.model.entity.base.MediaBase
import com.mxt.anitrend.util.media.MediaUtil
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.arrayContainingInAnyOrder
import org.hamcrest.Matchers.empty
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.`is`
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.reset
import org.mockito.junit.MockitoJUnitRunner
import java.util.ArrayList

@RunWith(MockitoJUnitRunner.StrictStubs::class)
class MediaUtilTests {

    @Mock
    private lateinit var list: MediaList

    @Mock
    private lateinit var media: MediaBase

    @Mock
    private lateinit var mediaTitle: MediaTitle

    @Before
    fun setupMocks() {
        `when`(list.media).thenReturn(media)
        `when`(media.title).thenReturn(mediaTitle)
    }

    @After
    fun resetMocks() {
        reset(list, media, mediaTitle)
    }

    @Test
    fun isAnimeType_givenNull_shouldReturnFalse() {
        assertThat(MediaUtil.isAnimeType(null), `is`(false))
    }

    @Test
    fun isAnimeType_givenAnime_shouldReturnTrue() {
        `when`(media.type).thenReturn(KeyUtil.ANIME)
        assertThat(MediaUtil.isAnimeType(media), `is`(true))
    }

    @Test
    fun isAnimeType_givenManga_shouldReturnFalse() {
        `when`(media.type).thenReturn(KeyUtil.MANGA)
        assertThat(MediaUtil.isAnimeType(media), `is`(false))
    }

    @Test
    fun isMangaType_givenNull_shouldReturnFalse() {
        assertThat(MediaUtil.isMangaType(null), `is`(false))
    }

    @Test
    fun isMangaType_givenAnime_shouldReturnFalse() {
        `when`(media.type).thenReturn(KeyUtil.ANIME)
        assertThat(MediaUtil.isMangaType(media), `is`(false))
    }

    @Test
    fun isMangaType_givenManga_shouldReturnTrue() {
        `when`(media.type).thenReturn(KeyUtil.MANGA)
        assertThat(MediaUtil.isMangaType(media), `is`(true))
    }

    @Test
    fun isIncrementLimitReached_ifProgressEqualToAnimeEpisodes_shouldReturnTrue() {
        val episodes = 10
        val progress = episodes

        `when`(media.type).thenReturn(KeyUtil.ANIME)
        `when`(list.progress).thenReturn(progress)
        `when`(media.episodes).thenReturn(episodes)

        assertThat(MediaUtil.isIncrementLimitReached(list), `is`(true))
    }

    @Test
    fun isIncrementLimitReached_ifProgressLessThanAnimeEpisodes_shouldReturnFalse() {
        val episodes = 10
        val progress = 3

        `when`(media.type).thenReturn(KeyUtil.ANIME)
        `when`(list.progress).thenReturn(progress)
        `when`(media.episodes).thenReturn(episodes)

        assertThat(MediaUtil.isIncrementLimitReached(list), `is`(false))
    }

    @Test
    fun isIncrementLimitReached_ifProgressGreaterThanAnimeEpisodes_shouldReturnFalse() {
        val episodes = 10
        val progress = 15

        `when`(media.type).thenReturn(KeyUtil.ANIME)
        `when`(list.progress).thenReturn(progress)
        `when`(media.episodes).thenReturn(episodes)

        assertThat(MediaUtil.isIncrementLimitReached(list), `is`(false))
    }

    @Test
    fun isIncrementLimitReached_ifProgressEqualToMangaChapters_shouldReturnTrue() {
        val chapters = 10
        val progress = chapters

        `when`(media.type).thenReturn(KeyUtil.MANGA)
        `when`(list.progress).thenReturn(progress)
        `when`(media.chapters).thenReturn(chapters)

        assertThat(MediaUtil.isIncrementLimitReached(list), `is`(true))
    }

    @Test
    fun isIncrementLimitReached_ifProgressLessThanMangaChapters_shouldReturnFalse() {
        val chapters = 10
        val progress = 7

        `when`(media.type).thenReturn(KeyUtil.MANGA)
        `when`(list.progress).thenReturn(progress)
        `when`(media.chapters).thenReturn(chapters)

        assertThat(MediaUtil.isIncrementLimitReached(list), `is`(false))
    }

    @Test
    fun isIncrementLimitReached_ifProgressGreaterThanMangaChapters_shouldReturnFalse() {
        val chapters = 10
        val progress = 20

        `when`(media.type).thenReturn(KeyUtil.MANGA)
        `when`(list.progress).thenReturn(progress)
        `when`(media.chapters).thenReturn(chapters)

        assertThat(MediaUtil.isIncrementLimitReached(list), `is`(false))
    }

    @Test
    fun isAllowedStatus_ifMediaIsNotYetReleased_shouldReturnFalse() {
        `when`(media.status).thenReturn(KeyUtil.NOT_YET_RELEASED)
        assertThat(MediaUtil.isAllowedStatus(list), `is`(false))
    }

    @Test
    fun isAllowedStatus_forAnyOtherStatus_shouldReturnTrue() {
        for (status in KeyUtil.MediaStatusValues.filterNotNull()) {
            if (KeyUtil.NOT_YET_RELEASED != status) {
                `when`(media.status).thenReturn(status)
                assertThat(
                    "Incrementing should be allowed for status: $status",
                    MediaUtil.isAllowedStatus(list),
                    `is`(true)
                )
            }
        }
    }

    @Test
    fun getMediaTitle_shouldReturnUserPreferredTitle() {
        val title = "Gintama"
        `when`(mediaTitle.userPreferred).thenReturn(title)
        assertThat(MediaUtil.getMediaTitle(media), equalTo(title))
    }

    @Test
    fun getMediaListTitle_shouldReturnUserPreferredTitle() {
        val title = "Gintama"
        `when`(mediaTitle.userPreferred).thenReturn(title)
        assertThat(MediaUtil.getMediaListTitle(list), equalTo(title))
    }

    @Test
    fun mapMediaTrend_shouldReturnCorrespondingMedia() {
        val media1 = mock(MediaBase::class.java)
        val media2 = mock(MediaBase::class.java)
        val media3 = mock(MediaBase::class.java)
        val mediaList = listOf(media1, media2, media3)

        val trendList = mediaList.map { media ->
            mock(MediaTrend::class.java).apply {
                `when`(this.media).thenReturn(media)
            }
        }

        assertThat(
            MediaUtil.mapMediaTrend(trendList).toTypedArray(),
            arrayContainingInAnyOrder(media1, media2, media3)
        )
    }

    @Test
    fun mapMediaTrend_givenNull_shouldReturnEmptyList() {
        assertThat(MediaUtil.mapMediaTrend(null), empty())
    }

    @Test
    fun getAiringMedia_shouldReturnReleasingMediaOnly() {
        val releasing = listOf(
            mock(MediaBase::class.java),
            mock(MediaBase::class.java)
        ).map { media ->
            mock(MediaList::class.java).apply {
                `when`(this.media).thenReturn(media)
            }
        }

        releasing.forEach { list ->
            `when`(list.media.status).thenReturn(KeyUtil.RELEASING)
        }

        val notReleasing = KeyUtil.MediaStatusValues
            .filterNotNull()
            .filter { it != KeyUtil.RELEASING }
            .map { status ->
                mock(MediaBase::class.java).apply {
                    `when`(this.status).thenReturn(status)
                }
            }
            .map { media ->
                mock(MediaList::class.java).apply {
                    `when`(this.media).thenReturn(media)
                }
            }

        val allMedia = ArrayList<MediaList>(releasing)
        allMedia.addAll(notReleasing)

        assertThat(
            MediaUtil.getAiringMedia(allMedia).toTypedArray(),
            equalTo(releasing.toTypedArray())
        )
    }

    @Test
    fun getAiringMedia_givenNull_shouldReturnEmptyList() {
        assertThat(MediaUtil.getAiringMedia(null), empty())
    }
}
